aten/src/ATen/ParallelOpenMP.h

```diff
@@ -30,15 +30,18 @@ inline void parallel_for(
   // returns false when omp_get_max_threads() == 1 inside nested "omp parallel"
   // See issue gh-32284
 
-#pragma omp parallel if (omp_get_max_threads() > 1 && !omp_in_parallel() && ((end - begin) > grain_size))
-  {
-    // choose number of tasks based on grain size and number of threads
-    // can't use num_threads clause due to bugs in GOMP's thread pool (See #32008)
```

Comment:
This comment alludes to the fact that using `num_threads` to choose dynamic number of threads causes a memory leak in some OpenMP implementations. This is the same implementation as before when the referenced issue was created.

Reply:
Thanks for your comment, @peterbell10!
@mattip had written [test/test_openmp.py](https://github.com/pytorch/pytorch/blob/master/test/test_openmp.py) in #36479 to check for memory leaks.

Reply:
@peterbell10, the proposed implementation is different from the pre-#32008 one in that it applies @mattip's recommendation in #32284 to prevent a memory leak:

https://github.com/pytorch/pytorch/issues/32284#issuecomment-612858186

I'll also test for the memory leak caused in #32008 & revert.

Reply:
>The key difference is that omp_in_parallel returns False in a parallel region if the number of OpenMP threads are 1, so the code accounts for it:

I assume you're talking about this line:
```cpp
int64_t num_threads = (omp_get_max_threads() > 1 && !omp_in_parallel()) ? omp_get_max_threads() : 1;
```
However, if `omp_get_max_threads()` returned 1 without the `> 1` check it doesn't matter what `omp_in_parallel()` returns, `num_threads` would be set to 1 in either branch.

Unless you're referring to something different?

>That test passes with the proposed change

The memory leak was specific to GNU's OpenMP implementation. It's possible that CI is using Intel OpenMP, or that the GNU bug might have been fixed in the mean time.

Reply:
> However, if `omp_get_max_threads()` returned 1 without the `> 1` check it doesn't matter what `omp_in_parallel()` returns, `num_threads` would be set to 1 in either branch.
> 
> Unless you're referring to something different?

In a parallel region with only 1 OpenMP thread, `omp_in_parallel()` returns 0.  Pre-#36479, it was causing an issue (double parallel for), so https://github.com/pytorch/pytorch/pull/36479/files changed the Pre-#36479 if clause to consider both `omp_get_max_threads() > 1` and `!omp_in_parallel()`, instead of only `!omp_in_parallel`.

In @mattip's own words,

> _omp_in_parallel() will only be true if we are already in a parallel context with more than one thread. But in the case at hand, we only have one thread, so omp_in_parallel() returns false, and we end up inside a double parallel for which somehow messes up OpenMP's memory management. I am not sure why that should leak memory, but if I modify the #pragma omp parallel in ParallelOpenMP.h to check num_threads > 1 then the leak goes away._

Reply:
> The memory leak was specific to GNU's OpenMP implementation. It's possible that CI is using Intel OpenMP, or that the GNU bug might have been fixed in the mean time.

Yes, I agree that it's possible that the libgomp bug might have been fixed.
However, what surprises me about #32008 is that @CaoZhongZ [reported that](https://github.com/pytorch/pytorch/issues/32008#issuecomment-578200133) all but `num_thread` threads, but the `num_threads` clause isn't supposed to cause the remaining threads in the OpenMP threadpool to exit.
In @CaoZhongZ's words,

> _The parallel region created by the code throttled number of OpenMP threads to 2 (num_threads) in our testcase. We could observe exiting of 126 threads at the end of the block, but MKL will boost the number of OpenMP threads back to 128, many threads spawning and exiting during the test and resulted (strongly suspected) the memory leak._ 

This part seems intriguing to me, so I'll try to investigate it on a machine with 32 physical cores.

Reply:
> In a parallel region with only 1 OpenMP thread, omp_in_parallel() returns 0. Pre-#36479, it was causing an issue (double parallel for), so #36479 (files) changed the Pre-#36479 if clause to consider both omp_get_max_threads() > 1 and !omp_in_parallel(), instead of only !omp_in_parallel.

Your tweak to `num_threads` assignment only matters when `omp_get_max_threads()` is 1, otherwise it's >1 so the condition is effectively just `!omp_in_parallel()` same as before. However, if `omp_get_max_threads()` is 1 then both outcomes of the ternary expression are 1. So, your tweak does nothing at all.

Reply:
I haven't tweaked anything. I used that from #36479 & the current code. I just stated that it's different from the pre-#32008 code.

Reply:
> However, if omp_get_max_threads() is 1 then both outcomes of the ternary expression are 1. So, your tweak does nothing at all.


In a parallel region, omp_get_max_threads, by default, is equal to the number of cores.
In a parallel region with one thread, `omp_in_parallel` would yield 0, so `!omp_in_parallel` would be 1.
Hence, both `!omp_in_parallel` & `omp_get_max_threads > 1` checks are required to ensure that we don't enter a double parallel for.

Reply:
Okay, forget about where the code comes from. Explain to me under what circumstances you expect `num_threads` to have a different value between these two expressions:

1. your version
```cpp
int64_t num_threads = (omp_get_max_threads() > 1 && !omp_in_parallel()) ? omp_get_max_threads() : 1;
```
2. the original version which lead to #32008
```cpp
int64_t num_threads = omp_in_parallel() ? 1 : omp_get_max_threads();
```

Reply:
Oops, you're absolutely right! I see now what you mean.

Reply:
However, in #32008, not using `omp_get_num_threads()` was fixing the memory-leak issue, so it seems like such a change wouldn't lead to similar issues.

Reply:
@peterbell10,

However, please look at the code which had caused #32008.
It's still different from this implementation.

The pre-#32008 code can enter double parallel-for.
This code won't.

Reply:
The only functional difference I can see between this and #26963 is the `if` clause. But that `if` clause doesn't have anything to do with nested parallel regions. So, why do you expect this to behave differently?

Reply:
I used an incorrect workaround for `omp_get_level()` :(

Reply:
Need to roll-back to previous commit, but it requires `omp_get_level()` support.